widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt

```diff
@@ -0,0 +1,328 @@
+        .subscribe { onQuickReplyReceived(it) }
+
+    private fun onMessage(): Flowable<State> = onMessageUseCase().toFlowable(BackpressureStrategy.BUFFER).withLatestFrom(state, ::mapNewMessage)
+
+    private fun onAction(): Flowable<State> = action.withLatestFrom(state, ::mapAction)
+
+    private fun checkUnsentMessages(state: State) {
+        sendUnsentMessagesUseCase(state.unsentItems.firstOrNull()?.message ?: return) {}
+    }
+
+    private fun mapChatHistory(historyResponse: ChatHistoryResponse): State {
+        if (historyResponse.items.isEmpty()) return State()
+
+        val chatItems: MutableList<ChatItem> = mutableListOf()
+        val chatItemIds: MutableSet<String> = mutableSetOf()
+
+        val rawItems = historyResponse.items
+
+
+        for (index in rawItems.indices.reversed()) {
+
+            val rawMessage = rawItems[index]
+            chatItemIds.add(rawMessage.chatMessage.id)
+
+            appendHistoryChatMessageUseCase(chatItems, rawMessage, index == rawItems.lastIndex)
+        }
+
+        chatItems.reverse()
+
+        if (addNewMessagesDividerUseCase(chatItems, historyResponse.newMessagesCount)) {
+            markMessagesReadWithDelay()
+        }
+
+        val lastMessageWithVisibleOperatorImage = chatItems.lastOrNull() as? OperatorChatItem
+
+        return State(
+            chatItems = chatItems,
+            chatItemIds = chatItemIds,
+            lastMessageWithVisibleOperatorImage = lastMessageWithVisibleOperatorImage
+        )
+    }
+
+    private fun mapNewMessage(chatMessage: ChatMessageInternal, messagesState: State): State {
+        if (messagesState.isNew(chatMessage)) {
+            appendNewChatMessageUseCase(messagesState, chatMessage)
+            if (chatMessage.chatMessage is VisitorMessage) {
+                checkUnsentMessages(messagesState)
+            }
+        }
+
+        return messagesState
+    }
+
+    private fun mapAction(action: Action, state: State): State {
+        return when (action) {
+            is Action.QueuingStarted -> mapInQueue(action.companyName, state)
+            is Action.OperatorConnected -> mapOperatorConnected(action, state)
+            Action.Transferring -> mapTransferring(state)
+            is Action.OperatorJoined -> mapOperatorJoined(action, state)
+            is Action.UnsentMessageReceived -> addUnsentMessage(action.message, state)
+            is Action.ResponseCardClicked -> mapResponseCardClicked(action.responseCard, state)
+            is Action.OnMediaUpgradeStarted -> mapMediaUpgrade(action.isVideo, state)
+            Action.OnMediaUpgradeToVideo -> mapUpgradeMediaToVideo(state)
+            Action.OnMediaUpgradeCanceled -> mapMediaUpgradeCanceled(state)
+            is Action.OnMediaUpgradeTimerUpdated -> mapMediaUpgradeTimerUpdated(action.formattedValue, state)
+            is Action.CustomCardClicked -> mapCustomCardClicked(action, state)
+            Action.ChatRestored -> state
```

Comment:
Why do some items start use `is` others - don't?

Reply:
Items without `is` are objects so no need to use the is keyword with them.

Reply:
Ah, got it, thanks!